buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionDownloadPlugin.java

```diff
@@ -0,0 +1,150 @@
+import org.elasticsearch.gradle.DistributionResolution;
+import org.elasticsearch.gradle.ElasticsearchDistribution;
+import org.elasticsearch.gradle.Version;
+import org.elasticsearch.gradle.VersionProperties;
+import org.elasticsearch.gradle.info.BuildParams;
+import org.elasticsearch.gradle.info.GlobalBuildInfoPlugin;
+import org.gradle.api.GradleException;
+import org.gradle.api.NamedDomainObjectContainer;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+
+import static org.elasticsearch.gradle.util.GradleUtils.projectDependency;
+
+/**
+ * An internal elasticsearch build plugin that registers additional
+ * distribution resolution strategies to the 'elasticsearch.download-distribution' plugin
+ * to resolve distributions from a local snapshot or a locally built bwc snapshot.
+ */
+public class InternalDistributionDownloadPlugin implements Plugin<Project> {
+
+    private BwcVersions bwcVersions = null;
+
+    @Override
+    public void apply(Project project) {
+        // this is needed for isInternal
+        project.getRootProject().getPluginManager().apply(GlobalBuildInfoPlugin.class);
+        if (!BuildParams.isInternal()) {
+            throw new GradleException(
+                "Plugin 'elasticsearch.internal-distribution-download' is not supported. "
+                    + "Use 'elasticsearch.distribution-download' plugin instead."
+            );
+        }
+        project.getPluginManager().apply(DistributionDownloadPlugin.class);
+        this.bwcVersions = BuildParams.getBwcVersions();
+        registerInternalDistributionResolutions(DistributionDownloadPlugin.getRegistrationsContainer(project));
+    }
+
+    /**
+     * Registers internal distribution resolutions.
+     *
+     * Elasticsearch distributions are resolved as project dependencies either representing
+     * the current version pointing to a project either under `:distribution:archives` or :distribution:packages`.
+     *
+     * BWC versions are resolved as project to projects under `:distribution:bwc`.
+     * */
+    private void registerInternalDistributionResolutions(NamedDomainObjectContainer<DistributionResolution> resolutions) {
+
+        resolutions.register("localBuild", distributionResolution -> distributionResolution.setResolver((project, distribution) -> {
+            if (VersionProperties.getElasticsearch().equals(distribution.getVersion())) {
+                // non-external project, so depend on local build
+                return projectDependency(project, distributionProjectPath(distribution), "default");
+            }
+            return null;
+        }));
+
+        resolutions.register("bwb", distributionResolution -> distributionResolution.setResolver((project, distribution) -> {
+            BwcVersions.UnreleasedVersionInfo unreleasedInfo = bwcVersions.unreleasedInfo(Version.fromString(distribution.getVersion()));
+            if (unreleasedInfo != null) {
+                assert distribution.getBundledJdk();
```

Comment:
Should we add a message here so that if folks incorrectly configure a snapshot bwc distribution w/o bundled JDK that we return an error as such?

Reply:
I changed this to throw a proper GradleException with an error message.